packages/oidc-dpop-client-browser/src/dpop/dpop.ts

```diff
@@ -30,14 +43,111 @@ import { JWK } from "node-jose";
  * @hidden
  */
 export async function generateJWK(
-  kty: "EC" | "OKP" | "RSA" | "oct",
+  kty: "EC" | "RSA",
   crvBitlength?: ECCurve | OKPCurve | number,
   parameters?: BasicParameters
 ): Promise<JSONWebKey> {
   const key = await JWK.createKey(kty, crvBitlength, parameters);
   return key.toJSON(true) as JSONWebKey;
 }
 
+/**
+ * Generates a Json Web Token (https://tools.ietf.org/html/rfc7519) containing
+ * the provided payload and using the signature algorithm specified in the options.
+ * @param payload The body of the JWT.
+ * @param key The key used for the signature.
+ * @param options
+ * @returns a 3-parts base64-encoded string, split by dots.
+ * @hidden
+ */
+export async function signJWT(
+  // eslint-disable-next-line @typescript-eslint/no-explicit-any
+  payload: Record<string, any>,
+  key: JWKECKey | JWKOKPKey | JWKRSAKey | JWKOctKey,
+  options?: JoseJWT.SignOptions
+): Promise<string> {
+  const parsedKey = await JWK.asKey(key);
+  const convertedKey: string = parsedKey.toPEM(true);
+  const signed = JWT.sign(payload, convertedKey, {
+    ...(options as JWT.SignOptions),
+  });
+  return signed;
+}
+
+/**
+ * Decodes the base64 Json Web Token into an object. If a key is specified, the
+ * JWT is also verified.
+ * @param token The base64-encoded token
+ * @param key The key used to sign the token
+ * @returns the payload of the JWT
+ * @hidden
+ */
+export async function decodeJWT(
+  token: string,
+  key?: JWKECKey | JWKOKPKey | JWKRSAKey | JWKOctKey,
+  options?: VerifyOptions
+): Promise<Record<string, unknown>> {
+  if (key) {
+    const parsedKey = await JWK.asKey(key);
+    const convertedKey: string = parsedKey.toPEM(false);
+    return JWT.verify(token, convertedKey, options) as Promise<
+      Record<string, unknown>
+    >;
+  }
+  return JWT.decode(token) as Promise<Record<string, unknown>>;
+}
+
+/**
+ * @param key
+ * @hidden
+ */
+export async function privateJWKToPublicJWK(
+  key: JSONWebKey
+): Promise<JSONWebKey> {
+  return (await JWK.asKey(key as JWK.RawKey, "public")) as JSONWebKey;
+}
+
+/**
+ * Normalizes a URL in order to generate the DPoP token based on a consistent scheme.
+ * @param audience The URL to normalize.
+ * @returns The normalized URL as a string.
+ * @hidden
+ */
+export function normalizeHtu(audience: URL): string {
+  return `${audience.origin}${audience.pathname}`;
+}
+
+/**
+ * Creates a DPoP header according to https://tools.ietf.org/html/draft-fett-oauth-dpop-04,
+ * based on the target URL and method, using the provided key.
+ * @param audience Target URL.
+ * @param method HTTP method allowed.
+ * @param key Key used to sign the token.
+ * @returns A JWT that can be used as a DPoP Authorization header.
+ */
+export async function createHeaderToken(
```

Comment:
Again, without a Class to provide context here, is this function better named as `createDpopHeaderToken()` (given the `typ` value here, it does seem to be very DPoP-specific)?
Raises another point on the function naming across all this code - we seem to be using `function xxxxJkt()` in places, and `function xxxToken()` in others - should we just be consistent everywhere (I'd suggest we use `Jkt` instead of `Token`, as it's more specific and explicit)...?

Reply:
...and likewise, I think we use `function xxxKey()` in places and `functionJwk()` in others (again I'd favor the more explicit `Jwk` instead of `Key`).

Reply:
I refactored the function name to `createDpopHeader` in f41a097f64666f7e014dc9b118009514961b9bf9, since it is indeed very DPoP-specific.

Reply:
And the change from `key` to `jwk` is in d411c4359f6f201b3c216821106daf542fb507e4